systems/roadbuilding.go

```diff
@@ -0,0 +1,414 @@
+}
+
+type RoadComponent struct {
+	From, To ecs.BasicEntity
+	Lanes    []*Lane
+
+	isHovered bool
+}
+
+type Lane struct {
+	ecs.BasicEntity
+	LaneComponent
+}
+
+type LaneComponent struct {
+	Commuters []*Commuter
+	Index     int
+}
+
+func (l *LaneComponent) Remove(basic ecs.BasicEntity) {
+	index := -1
+	for i, comm := range l.Commuters {
+		if comm.ID() == basic.ID() {
+			index = i
+			break
+		}
+	}
+	if index >= 0 {
+		l.Commuters = append(l.Commuters[:index], l.Commuters[index+1:]...)
+	}
+}
+
+type Commuter struct {
+	ecs.BasicEntity
+	common.SpaceComponent
+	common.RenderComponent
+	CommuterComponent
+}
+
+type CommuterComponent struct {
+	DistanceTravelled float32
+	DepartureTimes    []time.Duration
+	LastDeparture     int
+
+	Vehicle           Vehicle
+	Speed             float32
+	PreferredSpeed    float32
+	BrakeSpeed        float32
+	AccelerationSpeed float32
+
+	SwitchingLane     bool
+	SwitchingProgress float32
+
+	Road    commuterEntityRoad
+	Lane    *Lane
+	NewLane *Lane
+
+	// TODO: stuff like reaction time, amount of people,
+}
+
+type roadBuildingEntity struct {
+	*ecs.BasicEntity
+	*CityComponent
+	*common.SpaceComponent
+	*common.RenderComponent
+	*common.MouseComponent
+}
+
+type roadBuildingMoney struct {
+	*ecs.BasicEntity
+	*MoneyComponent
+}
+
+type RoadBuildingSystem struct {
+	world  *ecs.World
+	cities []roadBuildingEntity
+	money  roadBuildingMoney
+
+	roadHint       Road
+	roadCostHint   ui.Label
+	selectedEntity roadBuildingEntity
+	hovering       bool
+	mouseTracker   MouseTracker
+}
+
+func (r *RoadBuildingSystem) Remove(basic ecs.BasicEntity) {
+	delete := -1
+	for index, e := range r.cities {
+		if e.BasicEntity.ID() == basic.ID() {
+			delete = index
+			break
+		}
+	}
+	if delete >= 0 {
+		r.cities = append(r.cities[:delete], r.cities[delete+1:]...)
+	}
+
+	if basic.ID() == r.money.ID() {
+		r.money = roadBuildingMoney{}
+	}
+}
+
+func (r *RoadBuildingSystem) AddCity(basic *ecs.BasicEntity, city *CityComponent, space *common.SpaceComponent, render *common.RenderComponent, mouse *common.MouseComponent) {
+	r.cities = append(r.cities, roadBuildingEntity{basic, city, space, render, mouse})
+}
+
+func (r *RoadBuildingSystem) SetMoney(basic *ecs.BasicEntity, money *MoneyComponent) {
+	r.money = roadBuildingMoney{basic, money}
+}
+
+func (r *RoadBuildingSystem) New(w *ecs.World) {
+	r.world = w
+
+	r.mouseTracker.BasicEntity = ecs.NewBasic()
+	r.mouseTracker.MouseComponent = common.MouseComponent{Track: true}
+
+	fnt := &common.Font{
+		URL:  "fonts/Roboto-Regular.ttf",
+		FG:   color.RGBA{100, 0, 0, 200}, // dark red, but somewhat transparent
+		Size: 16,
+	}
+	err := fnt.CreatePreloaded()
+	if err != nil {
+		panic(err)
+	}
+
+	r.roadCostHint.Font = fnt
+
+	for _, system := range w.Systems() {
+		switch sys := system.(type) {
+		case *common.MouseSystem:
+			sys.Add(&r.mouseTracker.BasicEntity, &r.mouseTracker.MouseComponent, nil, nil)
+		}
+	}
+}
+
+func (r *RoadBuildingSystem) Update(dt float32) {
```

Comment:
This function is a doozy, trying to break this down/make less intimidating will be a big plus considering people will use this as a learning resource.